torch/nn/modules/module.py

```diff
@@ -770,22 +770,83 @@ def set_submodule(self, target: str, module: "Module") -> None:
             raise ValueError("Cannot set the submodule without a target name!")
 
         atoms: List[str] = target.split(".")
-        name = atoms.pop(-1)
-        mod: torch.nn.Module = self
+        if not isinstance(module, torch.nn.Module):
+            raise AttributeError("`" + atoms[-1] + "` is not an nn.Module")
+        if len(atoms) == 1:
+            parent: torch.nn.Module = self
+        else:
+            parent_key = ".".join(atoms[:-1])
+            parent = self.get_submodule(parent_key)
+        if not hasattr(parent, atoms[-1]):
+            raise AttributeError(
+                parent._get_name() + " has no attribute `" + atoms[-1] + "`"
+            )
+        mod = getattr(parent, atoms[-1])
+        if not isinstance(mod, torch.nn.Module):
+            raise AttributeError("`" + atoms[-1] + "` is not an nn.Module")
 
-        for item in atoms:
-            if not hasattr(mod, item):
-                raise AttributeError(
-                    mod._get_name() + " has no attribute `" + item + "`"
+        setattr(parent, atoms[-1], module)
+
+    def replace_or_create_new_leaf_module(self, target: str, module: "Module"):
```

Comment:
Why do we add a new API here instead of updating the documentation of `set_submodule` to reflect its actual behavior (of setting the submodule regardless of whether it exists)?

Reply:
Currently, `set_submodule` sets a module to a potentially non-existent attribute name only if the `target` is a non-dot-delimited string. This violates the docstring which says `Set the submodule given by target if it exists, otherwise throw an error.`

My understanding is that the docstring was the feature you guys wanted to support and the current (inconsistent) behaviour (depending on whether the `target` string is dot-delimited) is the bug.

You seem to be suggesting that the bug is a feature.

>Why do we add a new API

I saw that the undocumented behaviour (some may call it a bug) is used in the repository, suggesting the need for a method that adds a leaf node if it doesn't exist.

Reply:
Admittedly, the current docstring (released on the 2.5 stable release) describes the feature I was personally looking for, so I tested it out when I saw it. That's why I thought the behaviour described above is a bug.

I agree, this is a slightly strange PR in that it is named bugfix - and yet also introduces a new API/feature. As mentioned above, I thought this was necessary due to the use of `set_submodule` in this repository (see changed files in this PR) in a way that is inconsistent with its docstring (as per 2.5 stable release https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.set_submodule).

Please advise on steps forward.

Reply:
Got it, thank you! I missed the part where it's only possible to set a nonexistent submodule if the `target` string is not dot-delimited

> the undocumented behaviour (some may call it a bug) is used in the repository,

Are the callsites above actually using the undocumented behavior? It appears to me that they are all attempting to set submodules that exist, but do correct me if I missed something.

If so I would suggest just correcting `set_submodule` to properly throw an error in that one bad case

Reply:
> If so I would suggest just correcting set_submodule to properly throw an error in that one bad (non dot delimited) case and not adding a new replace_or_create_new_leaf_module API

That sounds good. From memory I tested using the recommended way e.g.,

```bash
python test/export/test_export.py TestExport.test_preserve_module_call_signature_unflatten_specialization
```
this line will break after the bugfix of `set_submodule`

https://github.com/pytorch/pytorch/pull/143455/files#diff-4a060a24e1f81389eab7390d434dddf919af50aa8fda6cbd81e182a53bd9328eL8473

I shall retest and let you know. If we indeed don't encounter problems with the `test_export.py`, it makes sense to remove the new api and confine the PR to the bugfix of `set_submodule`.

Reply:
I reverted the changes to `torch/export/unflatten.py` and `test/export/test_export.py` (reverted to using the now fixed `set_submodule` instead of the new `replace_or_create_new_leaf_module` api)

result of 

```bash
python ../test/export/test_export.py TestExport.test_unflatten_no_unroll
```

```
ERROR: test_unflatten_no_unroll (__main__.TestExport.test_unflatten_no_unroll)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/focal/coding/pytorch/torch/testing/_internal/common_utils.py", line 3108, in wrapper
    method(*args, **kwargs)
  File "/home/focal/coding/pytorch/torch/../test/export/test_export.py", line 8430, in test_unflatten_no_unroll
    test(
  File "/home/focal/coding/pytorch/torch/../test/export/test_export.py", line 8416, in test
    ufm = torch.export.unflatten(ep)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/focal/coding/pytorch/torch/export/unflatten.py", line 695, in unflatten
    return UnflattenedModule(module, flat_args_adapter)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/focal/coding/pytorch/torch/export/unflatten.py", line 480, in __init__
    self._dispatch_modules(redirected_call_indices, consts_targets)
  File "/home/focal/coding/pytorch/torch/export/unflatten.py", line 632, in _dispatch_modules
    self.set_submodule(
  File "/home/focal/coding/pytorch/torch/nn/modules/module.py", line 781, in set_submodule
    raise AttributeError(
AttributeError: UnflattenedModule has no attribute `n`

To execute this test, run the following from the base repo dir:
    python test/export/test_export.py TestExport.test_unflatten_no_unroll

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0

----------------------------------------------------------------------
Ran 1 test in 1.018s

FAILED (errors=1)
```

Reply:
ok, so I reverted the changes to the tests in `test/export/test_export.py` to use `set_submodule` again, but it does seem necessary to use `replace_or_create_new_leaf_module` in `torch/export/unflatten.py` in place of `set_submodule` in order for the tests in `test/export/test_export.py` to pass.

I tested with:

```
python ../test/export/test_export.py TestExport.test_unflatten_no_unroll
python ../test/export/test_export.py TestExport.test_preserve_module_call_signature_unflatten_specialization
python ../test/export/test_export.py TestExport.test_unflatten_multiple_graphs_preserve_signature_no_error
python ../test/export/test_export.py TestExport.test_unflatten_multiple_graphs_state
```

and everything runs without errors.

Reply:
Apologies for the delay, I would propose the following

- let's not add the new public API `replace_or_create_new_leaf_module`
- instead, `set_submodule` can take in a kwarg `strict` (similar to `strict` in `load_state_dict`)

if `strict=False`, do what `replace_or_create_new_leaf_module` does --
Replace the submodule given by ``target`` if it exists, or see if the parent module of target exists and create target, otherwise throw an error.

If `strict=True` use the changes to `set_submodule` with your bugfix -- 
Set the submodule given by ``target`` if it exists, otherwise throw an error

We can pass this kwarg as appropriate in the export logic